include/mitsuba/core/distr_1d.h

```diff
@@ -209,7 +214,26 @@ template <typename Value> struct DiscreteDistribution {
     }
 
 private:
-    void compute_cdf(const ScalarFloat *pmf, size_t size) {
+    void compute_cdf() {
+        static_assert(dr::is_jit_v<Float>,
+                      "This method should only be used in JIT variants!");
+
+        if (m_pmf.empty())
+            Throw("DiscreteDistribution: empty distribution!");
+        if (!dr::all(m_pmf >= 0.f))
+            Throw("DiscreteDistribution: entries must be non-negative!");
+        if (!dr::any(m_pmf > 0.f))
+            Throw("DiscreteDistribution: no probability mass found!");
+
+        m_cdf = dr::prefix_sum(m_pmf, false);
+        m_valid = ScalarVector2u(0, m_pmf.size() - 1);
+        m_sum = dr::gather<Float>(m_cdf, m_pmf.size() - 1);
+        m_normalization = 1.0 / m_sum;
+
+        dr::make_opaque(m_pmf, m_cdf, m_sum, m_normalization);
```

Comment:
make_opaque should not be needed here. It would only be non-opaque when the entire PMF is uniform, which is probably a nice special case where a specialized kernel is helpful.